rfcs/20230118-f16-accumulation/README.md

```diff
@@ -0,0 +1,125 @@
+Proposal for setting accumulation datatype.
+
+# 1. Introduction
+
+In order to maintain good accuracy in DNN workloads, it is important
+to use wide-enough accumulators during computation.  The default
+accumulation datatype in oneDNN is f32, for floating point
+computations, and s32 for integer computation.
+
+Some platforms allow additional speedups when using smaller
+accumulation datatype.  This is in particular the case for f16
+primitives, where up to 2x speedups can be obtained by using f16
+accumulation datatype on some platforms.
+
+# 2. Problem statement
+
+As of now, oneDNN officially supports only inference for f16 models,
+and many data show that good levels of accuracy can be maintained
+while using f16 accumulation datatype for most f16 models. We hence
+have been dispatching implementations with f16 accumulation on some
+platforms.
+
+Even though dispatching implementations with f16 accumulation is
+acceptable for inference, it is not for training, which requires the
+full f32 precision accumulators. Hence, as oneDNN is planning
+support for f16 training, it is important that we dispatch only
+implementation with f32 accumulation for training workloads.
+
+Note that the main user for oneDNN f16 implementations is openVINO,
+which is the requester for the implementations with f16 accumulation
+in the first place.
+
+
+# 3. Options
+
+## Using the propagation kind
+
+One natural option would be to rely on propagation kind.  If the
+propagation kind, is `forward_inference`, we allow dispatching of
+implementations with f16 accumulation, otherwise we don't.  This works
+for most primitives except for Matmul, which does not have a
+propagation kind. Hence this option does not cover the whole library.
+
+## Using the `ONEDNN_ENABLE_WORKLOAD` build option
+
+oneDNN currently has a `ONEDNN_ENABLE_WORKLOAD` build option that can
+be set to `INFERENCE` or `TRAINING` (default). This option is
+currently used to disable dispatching of some primitive
+implementations based on propagation_kind. We could rely on this
+mecanism to guard dispatching of implementations using f16
+accumulation with the two following caveats:
+- this would increase validation costs as we would need to validate both builds
+- it adds development burden. Currently we don't know any information
+  on the accumulation datatype for each implementations. Properly
+  guarding implementation dispatch would either need to be scattered
+  in implementation using f16 accumulation (which is brittle), or
+  would require to add internal APIs to query implementations for their
+  accumulation datatype.
+
+## A new `accumulation_datatype` attribute (recommended)
```

Comment:
We have global setting API for fpmath mode, do we expect to have similar API for this new attribute or is it required to set for each primitive?